quipucords/scanner/network/inspect_callback.py

```diff
@@ -1,239 +1,298 @@
+
+
+@dataclass
+class AnsibleResults:
+    """Host state and facts collected from ansible events."""
 
+    host: str
+    status: str
+    facts: dict
 
-class InspectResultCallback:
-    """A sample callback plugin used for performing an action.
 
-    If you want to collect all results into a single object for processing at
-    the end of the execution, look into utilizing the ``json`` callback plugin
-    or writing your own custom callback plugin
-    """
+class InspectCallback:
+    """Callback helper to plug ansible callbacks for network scan inspection phase."""
 
-    def __init__(self, scan_task, manager_interrupt):
-        """Create result callback."""
-        self.scan_task = scan_task
-        self.source = scan_task.source
-        self._ansible_facts = {}
-        self.last_role = None
+    def __init__(self, manager_interrupt):
+        self._ansible_facts = defaultdict(dict)
+        self._unreachable_hosts = set()
         self.stopped = False
         self.interrupt = manager_interrupt
 
-    def process_task_facts(self, task_facts, host):
-        """Collect, process, and save task facts."""
-        host_facts = {}
-        for key, value in task_facts.items():
-            if key == HOST_DONE:
-                self._finalize_host(host, SystemInspectionResult.SUCCESS)
+    def iter_results(self) -> Generator[AnsibleResults]:
+        """Yield host completion state and ansible facts."""
+        for host, facts in self._ansible_facts.items():
+            if host in self._unreachable_hosts:
+                host_status = SystemInspectionResult.UNREACHABLE
+            elif facts.get(HOST_DONE, False) is True:
+                # host_done is the last fact - we assume a host is successfully scanned
+                # if this fact is set to true
+                host_status = SystemInspectionResult.SUCCESS
             else:
-                processed_value = process.process(
-                    self.scan_task, self._ansible_facts.get(host, {}), key, value, host
-                )
-                host_facts[key] = processed_value
-        if bool(host_facts) and host != UNKNOWN:
-            if host in self._ansible_facts:
-                self._ansible_facts[host].update(host_facts)
-            else:
-                self._ansible_facts[host] = host_facts
+                host_status = SystemInspectionResult.FAILED
+            yield AnsibleResults(host=host, status=host_status, facts=facts)
 
-    def task_on_ok(self, event_dict):
-        """Print a json representation of the event_data on ok."""
-        event_data = event_dict.get("event_data")
-        host = event_data.get("host", UNKNOWN)
-        result = event_data.get("res")
-        task_action = event_data.get("task_action")
+    def task_on_ok(self, event_dict: dict):
+        """Handle successful ansible events."""
+        # example event
+        # {
+        #   "uuid": "9d097760-4d2c-43e5-8aba-d1a0e825aed7",
+        #   "counter": 140,
```

Comment:
I failed miserably in finding documentation for those events, so I decided to slap full examples here.
they will be used later (with way less data) in unit tests for this class

Reply:
In this case, I'd encourage including long example objects like this one _in full_ in the unit tests (maybe with a comment to explain how you captured them). Although we might think we need less data for the tests, I often look at tests as another form of documentation, _especially_ when it comes to dealing with external/third-party integration code, and these fuller examples might contain clues for future ideas.

Reply:
I moved all these comments to actual fixtures with instructions on how to run obtain them